pisa/stages/osc/pi_2nu.py

```diff
@@ -0,0 +1,147 @@
+
+    None
+
+    Notes
+    -----
+
+    """
+    def __init__(self,
+                 data=None,
+                 params=None,
+                 input_names=None,
+                 output_names=None,
+                 debug_mode=None,
+                 input_specs=None,
+                 calc_specs=None,
+                 output_specs=None,
+                ):
+
+        expected_params = (
+                           'theta23',
+                           'deltam31',
+                          )
+
+        input_names = ()
+        output_names = ()
+
+        # what are the keys used from the inputs during apply
+        input_apply_keys = ('weights',
+                            'nu_flux',
+                            'true_energy',
+                            'true_coszen',
+                           )
+
+        # what keys are added or altered for the outputs during apply
+        output_apply_keys = ('weights',
+                      )
+
+        # init base class
+        super(pi_2nu, self).__init__(
+            data=data,
+            params=params,
+            expected_params=expected_params,
+            input_names=input_names,
+            output_names=output_names,
+            debug_mode=debug_mode,
+            input_specs=input_specs,
+            calc_specs=calc_specs,
+            output_specs=output_specs,
+            input_apply_keys=input_apply_keys,
+            output_apply_keys=output_apply_keys,
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is not None
+        assert self.output_mode is not None
+
+    @profile
+    def apply_function(self):
+
+        theta23 = self.params.theta23.value.m_as('rad')
+        deltam31 = self.params.deltam31.value.m_as('eV**2')
+
+        for container in self.data:
+            if 'numu' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    theta23,
+                                    deltam31,
+                                    container['true_energy'].get(WHERE),
+                                    container['true_coszen'].get(WHERE),
+                                    1,
+                                    out=container['weights'].get(WHERE),
+                                   )
+
+            if 'nutau' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    theta23,
+                                    deltam31,
+                                    container['true_energy'].get(WHERE),
+                                    container['true_coszen'].get(WHERE),
+                                    3,
+                                    out=container['weights'].get(WHERE),
+                                   )
+            if 'nue' in container.name:
+              apply_probs_vectorized(container['nu_flux'].get(WHERE),
+                                    theta23,
```

Comment:
put `FTYPE()` around these extra values that are not arrays, just to make sure they are casted to the right precision